### PR TITLE
Added unit tests to verify the observers thread-safety.

### PR DESCRIPTION
**Pull request by sgl0v**

### CODE REVIEW

File: PMKVObserverTests/PMKVObserverTests.m [hunk 1]
```diff
@@ -239,4 +239,67 @@ - (void)testInitialCancel {
...
+    NSString* value1 = @"value1";
+    NSString* value2 = @"value2";
+
+    PMKVObserver *token = [PMKVObserver observeObject:self.helper keyPath:@"str" options:NSKeyValueObservingOptionNew block:^(id  _Nonnull object, NSDictionary<NSString *,id> * _Nullable change, PMKVObserver * _Nonnull kvo) {
+        XCTAssertEqualObjects([object str], value1);
+        fired = YES;
+    }];
+    dispatch_sync(queue, ^{
+        self.helper.str = value1;
+        [token cancel];
```
Comment (lilyball) [inline PMKVObserverTests/PMKVObserverTests.m:254]: This isn't actually doing anything on a background thread at all. `dispatch_sync()` always executes the block on the current thread unless the queue being disaptched to is the main queue (and the process is using `NSRunLoop` or `CFRunLoop`; a process using `dispatch_main` doesn't tie the main queue to thread 0).

File: PMKVObserverTests/PMKVObserverTests.m [hunk 2]
```diff
@@ -239,4 +239,67 @@ - (void)testInitialCancel {
...
+
+    __block NSObject* observer = [[NSObject alloc] init];
+    [PMKVObserver observeObject:self.helper observer:observer keyPath:@"str" options:NSKeyValueObservingOptionNew block:^(id  _Nonnull observer, id  _Nonnull object, NSDictionary<NSString *,id> * _Nullable change, PMKVObserver * _Nonnull kvo) {
+        XCTAssertEqualObjects([object str], value1);
+        fired = YES;
+    }];
+
+    dispatch_sync(queue, ^{
+        self.helper.str = value1;
+        observer = nil;
```
Comment (lilyball) [inline PMKVObserverTests/PMKVObserverTests.m:275]: Same here, this isn't in fact on a background thread.

File: PMKVObserverTests/PMKVObserverTests.m [hunk 3]
```diff
@@ -239,4 +239,67 @@ - (void)testInitialCancel {
...
+    __block PMKVObserver *token = nil;
+    dispatch_sync(queue1, ^{
+        token = [PMKVObserver observeObject:self.helper observer:self keyPath:@"str" options:NSKeyValueObservingOptionNew block:^(id  _Nonnull observer, id  _Nonnull object, NSDictionary<NSString *,id> * _Nullable change, PMKVObserver * _Nonnull kvo) {
+            XCTAssertEqualObjects([object str], value1);
+            fired = YES;
+        }];
+    });
+    self.helper.str = value1;
+    dispatch_sync(queue2, ^{
+        [token cancel];
```
Comment (lilyball) [inline PMKVObserverTests/PMKVObserverTests.m:298]: Nor is this.

**Comment by lilyball** [comment]
I appreciate the PR, but even if you change the `dispatch_sync` to use `dispatch_async` + a semaphore or group or `dispatch_block_wait`, it's not really testing anything at all other than the fact that the library does not assert that it's on the main thread. Even were the library not thread-safe, your test here wouldn't be expected to reveal any issues, because there's no actual _race_ going on. The only way to actually test the thread safety is to reliably create a lot of races and make sure that it behaves correctly every single time, and it's not very practical to do that.

**Comment by sgl0v** [comment]
Yeah, these tests don't bring big value, just test the expected behaviour: KVO wrapper works in the same way with one or multiple threads ("Observers can be registered on a different thread than KVO notifications are sent on, and can be cancelled on yet another thread. An observer can even be cancelled from two thread simultaneously"). As you said, to make proper tests for thread safety we should create a lot of races conditions and check that it behaves correctly.

File: PMKVObserverTests/PMKVObserverTests.m (hunk 1 above)
Comment (sgl0v) [inline PMKVObserverTests/PMKVObserverTests.m:254]: I totally forgot this `dispatch_sync` issue. According to the documentation: "As an optimization, this function invokes the block on the current thread when possible." .

**Comment by lilyball** [comment]
Those claims aren't very interesting without data races (well, being able to cancel twice is interesting regardless of threading, but that's not what you're trying to test here). We already know that KVO supports registering and deregistering on different threads, even a library that doesn't attempt to address thread safety at all would get that sort of basic functionality for free. It's proper handling of data races that's interesting.

**Comment by lilyball** [comment]
FWIW, based on this conversion I went ahead and pushed some tests yesterday that ensure that you can call `-cancel` twice in a row without a problem.

**Pull request closed without merging**